### PR TITLE
[WIP]crimson/osd/osd_operations/pg_advance: make PGAdvanceMap a pg_operation

### PR DESCRIPTION
**Pull request by Matan-B**
[eb27130aa424e65738342b3baf26d52e21241ab2] Fixes the following case:
```
DEBUG 2023-05-09 13:27:22,883 [shard 0] osd - pg_advance_map(id=13033, detail=PGAdvanceMap(pg=10.16 from=76 to=77)): complete
DEBUG 2023-05-09 13:27:22,883 [shard 0] osd - pg_advance_map(id=13152, detail=PGAdvanceMap(pg=15.11 from=76 to=77)): sending pg temp
DEBUG 2023-05-09 13:27:22,883 [shard 0] osd - pg_advance_map(id=12506, detail=PGAdvanceMap(pg=3.1b from=77 to=76)): start: getting map 78
DEBUG 2023-05-09 13:27:22,883 [shard 0] osd - get_local_map loading osdmap.78 from disk
INFO  2023-05-09 13:27:22,883 [shard 0] osd - load_map osdmap.78
DEBUG 2023-05-09 13:27:22,883 [shard 0] osd - load_map_bl loading osdmap.78 from disk
DEBUG 2023-05-09 13:27:22,883 [shard 0] alienstore - read
...
DEBUG 2023-05-09 13:27:22,957 [shard 0] osd - pg_advance_map(id=13074, detail=PGAdvanceMap(pg=12.d from=76 to=77)): complete
DEBUG 2023-05-09 13:27:22,957 [shard 0] osd - pg_advance_map(id=12975, detail=PGAdvanceMap(pg=8.b from=76 to=77)): sending pg temp
DEBUG 2023-05-09 13:27:22,957 [shard 0] osd - pg_advance_map(id=13144, detail=PGAdvanceMap(pg=15.7 from=76 to=77)): complete
DEBUG 2023-05-09 13:27:22,957 [shard 0] osd - pg_advance_map(id=12833, detail=PGAdvanceMap(pg=2.2 from=76 to=77)): sending pg temp
DEBUG 2023-05-09 13:27:22,957 [shard 0] osd - OSDMeta::load_map: start read gave enoent on 78
ERROR 2023-05-09 13:27:22,957 [shard 0] none - ../src/crimson/osd/osd_meta.cc:40 : In function 'OSDMeta::load_map(epoch_t)::<lambda()>', abort(%s)
abort() called
```
```
// See PGAdvanceMap(pg=3.1b from=77 to=76)
// from is set via `pg->get_osdmap_epoch()`
// 'from' is already newer than 'to'
// we start iterating from 77+1 (osdmap.78 does not exist)
```

Drafted for discussion - comments left in commit. The fix assumes that if the pg was already advanced to a later osdmap than `to` - we can skip the `PGAdvanceMap` operation.
 
https://gist.github.com/Matan-B/67a6c2a9b9c581600799c2c37704e913 

Fixes: https://tracker.ceph.com/issues/59165

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by athanatos** [comment]
So the bug is that the PG already has a map epoch after to?  Is it that we're processing the same maps from multiple MOSDMap messages?  It's entirely possible, likely even, for an OSD to get several MOSDMap messages with overlapping map sequences close together since any peer OSD will send one if it thinks the OSD is out of date.  Likely means we're not correctly maintaining the state that would avoid processing the same map more than once.

**Review by athanatos — CHANGES_REQUESTED** [review]
(See question above)

**Comment by Matan-B** [comment]
> So the bug is that the PG already has a map epoch after to? Is it that we're processing the same maps from multiple MOSDMap messages? It's entirely possible, likely even, for an OSD to get several MOSDMap messages with overlapping map sequences close together since any peer OSD will send one if it thinks the OSD is out of date. Likely means we're not correctly maintaining the state that would avoid processing the same map more than once.

@athanatos, I think this is caused by misusing Seastar's `do_for_each` implementation.
The only check is for `begin != end` which assumes that `begin` iterator is smaller than `end` iterator [1].
This can also be resolved by using `do_until` with a safer stop condition.

I will also note that in an early test run, it seems like this issue no longer appear with 56/60 passed [2]

[1]https://github.com/scylladb/seastar/blob/68b5aece19d1b3656a95c6869ff510db2e402dec/include/seastar/core/loop.hh#L424

[2]https://pulpito.ceph.com/matan-2023-05-11_08:22:25-crimson-rados-crimson-only-no-mclock-osdmap-enoent-distro-crimson-smithi/

**Comment by athanatos** [comment]
It's not so much that it's a misuse as that the code as written encodes an assumption that we don't initiate PGAdvanceMap on the same epoch twice.  It seems to me like we really do want to enforce that invariant, so I'd like to understand how the PG advanced past that epoch in the first place.  Tolerating that case seems at best inefficient and at worst would hide a problem.

**Review by athanatos — CHANGES_REQUESTED** [review]
(See above)

**Comment by Matan-B** [comment]
> It's not so much that it's a misuse as that the code as written encodes an assumption that we don't initiate PGAdvanceMap on the same epoch twice. It seems to me like we really do want to enforce that invariant, so I'd like to understand how the PG advanced past that epoch in the first place. Tolerating that case seems at best inefficient and at worst would hide a problem.

I have made some remarks based on a more thorough examination of this issue and provided an abstract solution/re-design to the current way we schedule a pg advance map operation.
Can you please take a look at https://gist.github.com/Matan-B/67a6c2a9b9c581600799c2c37704e913 for any initial feedback?
Thank you.

**Review by athanatos — CHANGES_REQUESTED** [review]
@Matan-B IIRC, the plan here was to add a mutex or some other mechanism to correctly pipeline concurrent MOSDMap messages, right?

@athanatos Right, it's wip.

**Comment by Matan-B** [comment]
> @Matan-B IIRC, the plan here was to add a mutex or some other mechanism to correctly pipeline concurrent MOSDMap messages, right?


@athanatos Right, it's wip.

File: src/crimson/osd/pg_shard_manager.cc [hunk 1]
```diff
@@ -109,15 +110,24 @@ PGShardManager::get_pg_stats() const
...
+    auto &state = local_service.local_state;
+    auto &pgs = state.pg_map.get_pgs();
+    return seastar::parallel_for_each(
+      pgs.begin(), pgs.end(),
+      [this, conn, advance_from, advance_to, &state](auto& pg) {
+        return start_pg_operation_with_state<PGAdvanceMap>(
+          state,
+          conn,
+          shard_services.local(),
+          pg.second, advance_from, advance_to,
```
Comment (Matan-B) [inline src/crimson/osd/pg_shard_manager.cc:126]: This essentially will allow us to schedule a PGAdvanceMap as an exclusive pg_operation with bounded `advance_from` and `advance_to` epochs.

(Pushed for early round of review)

File: src/crimson/osd/osd.cc [hunk 2]
```diff
@@ -853,8 +853,58 @@ bool OSD::require_mon_peer(crimson::net::Connection *conn, Ref<Message> m)
   return true;
 }
 
-seastar::future<> OSD::handle_osd_map(crimson::net::ConnectionRef conn,
-                                      Ref<MOSDMap> m)
+interval_set<epoch_t> OSD::calc_new_osd_maps(
+  const epoch_t& origin_first,
```
Comment (ronen-fr) [inline src/crimson/osd/osd.cc:857]: epoch_t is a 32b integer. Better to pass it by value.

Comment (ronen-fr) [reply src/crimson/osd/osd.cc:857]: (true even if will be turned into a 64b value in the future...)

File: src/crimson/osd/osd.cc [hunk 3]
```diff
@@ -853,8 +853,58 @@ bool OSD::require_mon_peer(crimson::net::Connection *conn, Ref<Message> m)
...
+  const epoch_t& origin_last)
+{
+  interval_set<epoch_t> origin_epochs;
+  interval_set<epoch_t> old_epochs;
```
Comment (ronen-fr) [inline src/crimson/osd/osd.cc:861]: not needed until 8 lines below...

**Pull request closed without merging**